src/compute-client/src/controller.rs

```diff
@@ -942,46 +952,50 @@ pub struct CollectionState<T> {
+    /// holds through [`ReadHold`] objects acquired through [`Instance::acquire_read_hold`].
+    ///
+    /// TODO(teskje): Restructure the code to enforce the above in the type system.
     read_capabilities: MutableAntichain<T>,
-    /// The implicit capability associated with collection creation.
-    implied_capability: Antichain<T>,
-    /// A capability held to enable dataflow warmup.
+    /// A read hold maintaining the implicit capability associated with collection creation.
+    implied_read_hold: ReadHold<T>,
+    /// A read hold held to enable dataflow warmup.
     ///
     /// Dataflow warmup is an optimization that allows dataflows to immediately start hydrating
     /// even when their next output time (as implied by the `write_frontier`) is in the future.
     /// By installing a read capability derived from the write frontiers of the collection's
     /// inputs, we ensure that the as-of of new dataflows installed for the collection is at a time
     /// that is immediately available, so hydration can begin immediately too.
-    warmup_capability: Antichain<T>,
-    /// The policy to use to downgrade `self.implied_capability`.
+    warmup_read_hold: ReadHold<T>,
+    /// The policy to use to downgrade `self.implied_read_hold`.
     ///
     /// If `None`, the collection is a write-only collection (i.e. a sink). For write-only
-    /// collections, the `implied_capability` is only required for maintaining read holds on the
+    /// collections, the `implied_read_hold` is only required for maintaining read holds on the
     /// inputs, so we can immediately downgrade it to the `write_frontier`.
     read_policy: Option<ReadPolicy<T>>,
 
-    /// Storage identifiers on which this collection depends.
-    storage_dependencies: Vec<GlobalId>,
-    /// Compute identifiers on which this collection depends.
-    compute_dependencies: Vec<GlobalId>,
+    /// Storage identifiers on which this collection depends, and read holds this collection
+    /// requires on them.
+    storage_dependencies: BTreeMap<GlobalId, ReadHold<T>>,
+    /// Compute identifiers on which this collection depends, and read holds this collection
+    /// requires on them.
+    compute_dependencies: BTreeMap<GlobalId, ReadHold<T>>,
 
     /// The write frontier of this collection.
     write_frontier: Antichain<T>,
-    /// The write frontiers reported by individual replicas.
-    replica_write_frontiers: BTreeMap<ReplicaId, Antichain<T>>,
-    /// The input frontiers reported by individual replicas.
-    replica_input_frontiers: BTreeMap<ReplicaId, Antichain<T>>,
 
     /// Introspection state associated with this collection.
     collection_introspection: CollectionIntrospection<T>,
 }
 
-impl<T> CollectionState<T> {
+impl<T: Timestamp> CollectionState<T> {
     /// Reports the current read capability.
     pub fn read_capability(&self) -> &Antichain<T> {
-        &self.implied_capability
+        self.implied_read_hold.since()
```

Comment:
I think it's mainly named this because I didn't want to leak the "implied" thing to adapter code. It's bad enough that we have to know what it means :D The thinking is that this is the frontier at which you are capable of reading, i.e. at which you can still get a read hold. There is also `read_frontier` which is the actual read frontier of the collection, but that's only for writing it down to introspection relations, you are not guaranteed to get a read hold at this frontier. Yes, it's confusing.

>  Could we maybe rename it?

Sure, but finding a name is again difficult!

There is a follow-up I want to do that refactors adapter's `read_policy.rs` to only store compute and storage `ReadHold`s, rather than manually creating its own surrogates. I hope that if we do that, adapter doesn't need to call this `read_capability` method anymore and we can just delete it.